File: sdks/go/dcp/v2/error_codes.go
// Package v2 — DCP canonical error codes (Go port).

package v2

import "fmt"

// DcpErrorCode is a stable string identifier shared across SDKs.
type DcpErrorCode string

const (
	// Schema (E001-E099)
	ErrBundleSchemaInvalid   DcpErrorCode = "DCP-E001"
	ErrArtifactSchemaInvalid DcpErrorCode = "DCP-E002"
	ErrVersionUnsupported    DcpErrorCode = "DCP-E003"
	ErrManifestMissing       DcpErrorCode = "DCP-E004"

	// Signature (E100-E199)
	ErrClassicalSigInvalid    DcpErrorCode = "DCP-E100"
	ErrPqSigInvalid           DcpErrorCode = "DCP-E101"
	ErrCompositeBindingInvalid DcpErrorCode = "DCP-E102"
	ErrSignatureMissing        DcpErrorCode = "DCP-E103"
	ErrSignatureExpired        DcpErrorCode = "DCP-E104"

	// Hash/Chain (E200-E299)
	ErrHashChainBroken      DcpErrorCode = "DCP-E200"
	ErrManifestHashMismatch DcpErrorCode = "DCP-E201"
	ErrMerkleRootMismatch   DcpErrorCode = "DCP-E202"
	ErrDualHashMismatch     DcpErrorCode = "DCP-E203"
	ErrIntentHashMismatch   DcpErrorCode = "DCP-E204"

	// Identity (E300-E399)
	ErrAgentRevoked DcpErrorCode = "DCP-E300"
	ErrKeyExpired   DcpErrorCode = "DCP-E301"
	ErrKeyRevoked   DcpErrorCode = "DCP-E302"
	ErrKidMismatch  DcpErrorCode = "DCP-E303"
	ErrRprInvalid   DcpErrorCode = "DCP-E304"

	// Policy (E400-E499)
	ErrTierInsufficient DcpErrorCode = "DCP-E400"
	ErrPolicyViolation  DcpErrorCode = "DCP-E401"
	ErrDowngradeAttempt DcpErrorCode = "DCP-E402"
	ErrCapabilityDenied DcpErrorCode = "DCP-E403"

	// Session (E500-E599)
	ErrSessionNonceInvalid DcpErrorCode = "DCP-E500"
	ErrSessionExpired      DcpErrorCode = "DCP-E501"
	ErrSessionReplay       DcpErrorCode = "DCP-E502"
	ErrSequenceOutOfOrder  DcpErrorCode = "DCP-E503"

	// A2A (E600-E699)
	ErrA2AHandshakeFailed    DcpErrorCode = "DCP-E600"
	ErrA2ABundleRejected     DcpErrorCode = "DCP-E601"
	ErrA2ACapabilityMismatch DcpErrorCode = "DCP-E602"
	ErrA2ASessionClosed      DcpErrorCode = "DCP-E603"
	ErrA2ADecryptFailed      DcpErrorCode = "DCP-E604"

	// Rate limiting (E700-E799)
	ErrRateLimitExceeded DcpErrorCode = "DCP-E700"
	ErrCircuitOpen       DcpErrorCode = "DCP-E701"
	ErrBackpressure      DcpErrorCode = "DCP-E702"

	// Internal (E900-E999)
	ErrInternal             DcpErrorCode = "DCP-E900"
	ErrAlgorithmUnavailable DcpErrorCode = "DCP-E901"
	ErrHSM                  DcpErrorCode = "DCP-E902"
)

type errInfo struct {
	message   string
	retryable bool
}

var errorDescriptions = map[DcpErrorCode]errInfo{
	ErrBundleSchemaInvalid:     {"Bundle does not conform to DCP schema", false},
	ErrArtifactSchemaInvalid:   {"Artifact does not conform to DCP schema", false},
	ErrVersionUnsupported:      {"DCP version not supported", false},
	ErrManifestMissing:         {"Bundle manifest is missing", false},
	ErrClassicalSigInvalid:     {"Classical (Ed25519) signature verification failed", false},
	ErrPqSigInvalid:            {"Post-quantum signature verification failed", false},
	ErrCompositeBindingInvalid: {"Composite signature binding is invalid", false},
	ErrSignatureMissing:        {"Required signature is missing", false},
	ErrSignatureExpired:        {"Signature has expired", false},
	ErrHashChainBroken:         {"Audit hash chain integrity check failed", false},
	ErrManifestHashMismatch:    {"Manifest hash does not match artifact", false},
	ErrMerkleRootMismatch:      {"Merkle root does not match audit entries", false},
	ErrDualHashMismatch:        {"Dual hash chain inconsistency detected", false},
	ErrIntentHashMismatch:      {"Intent hash does not match", false},
	ErrAgentRevoked:            {"Agent has been revoked", false},
	ErrKeyExpired:              {"Signing key has expired", false},
	ErrKeyRevoked:              {"Signing key has been revoked", false},
	ErrKidMismatch:             {"Key identifier does not match public key", false},
	ErrRprInvalid:              {"Responsible Principal Record is invalid", false},
	ErrTierInsufficient:        {"Security tier does not meet minimum requirement", false},
	ErrPolicyViolation:         {"Action violates policy", false},
	ErrDowngradeAttempt:        {"Security tier downgrade is not allowed", false},
	ErrCapabilityDenied:        {"Requested capability is not authorized", false},
	ErrSessionNonceInvalid:     {"Session nonce is invalid", false},
	ErrSessionExpired:          {"Session has expired", false},
	ErrSessionReplay:           {"Session replay detected", false},
	ErrSequenceOutOfOrder:      {"Message sequence out of order", false},
	ErrA2AHandshakeFailed:      {"A2A handshake failed", true},
	ErrA2ABundleRejected:       {"Peer rejected presented bundle", false},
	ErrA2ACapabilityMismatch:   {"Peer does not satisfy requested capabilities", false},
	ErrA2ASessionClosed:        {"A2A session is closed", false},
	ErrA2ADecryptFailed:        {"AES-GCM decryption failed (tag mismatch)", false},
	ErrRateLimitExceeded:       {"Rate limit exceeded", true},
	ErrCircuitOpen:             {"Circuit breaker is open", true},
	ErrBackpressure:            {"Backpressure applied — retry later", true},
	ErrInternal:                {"Internal error", true},
	ErrAlgorithmUnavailable:    {"Requested algorithm is not registered", false},
	ErrHSM:                     {"Hardware security module reported an error", true},
}

// DcpError is the canonical DCP error struct.
type DcpError struct {
	Code      DcpErrorCode
	Message   string
	Retryable bool
	Timestamp string
	Details   map[string]interface{}
}

// Error implements the `error` interface so DcpError can be returned from funcs.
func (e *DcpError) Error() string {
	return fmt.Sprintf("[%s] %s", e.Code, e.Message)
}

// Retry returns the retryable flag for a code.
func (c DcpErrorCode) Retry() bool {
	info, ok := errorDescriptions[c]
	if !ok {
		return false
	}
	return info.retryable
}

// Description returns the canonical message + retryable flag.
func (c DcpErrorCode) Description() (string, bool) {
	info, ok := errorDescriptions[c]
	if !ok {
		return "Unknown error", false
	}
	return info.message, info.retryable
}

// CreateDcpError builds a DcpError with the canonical message + retryable flag.
// If `message` is "", the default description is used.
func CreateDcpError(code DcpErrorCode, message string, details map[string]interface{}) *DcpError {
	msg, retryable := code.Description()
	if message == "" {
		message = msg
	}
	if details == nil {
		details = map[string]interface{}{}
	}
	return &DcpError{
		Code:      code,
		Message:   message,
		Retryable: retryable,
		Timestamp: utcNowISO(),
		Details:   details,
	}
}

// IsDcpError reports whether err is a *DcpError.
func IsDcpError(err error) bool {
	_, ok := err.(*DcpError)
	return ok
}

// DetectWireFormat returns "json" or "cbor" based on the first byte.
func DetectWireFormat(data []byte) string {
	if len(data) == 0 {
		return "json"
	}
	switch data[0] {
	case 0x7B, 0x5B, 0x20, 0x0A, 0x0D, 0x09:
		return "json"
	}
	return "cbor"
}

File: sdks/rust/src/v2/error_codes.rs
//! DCP v2.0 Canonical error codes — Rust port.

use std::fmt;

/// DCP canonical error code. The display representation (e.g. `DCP-E100`)
/// is the stable identifier shared across SDKs.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum DcpErrorCode {
    // Schema errors (E001-E099)
    BundleSchemaInvalid,
    ArtifactSchemaInvalid,
    VersionUnsupported,
    ManifestMissing,
    // Signature errors (E100-E199)
    ClassicalSigInvalid,
    PqSigInvalid,
    CompositeBindingInvalid,
    SignatureMissing,
    SignatureExpired,
    // Hash/Chain errors (E200-E299)
    HashChainBroken,
    ManifestHashMismatch,
    MerkleRootMismatch,
    DualHashMismatch,
    IntentHashMismatch,
    // Identity errors (E300-E399)
    AgentRevoked,
    KeyExpired,
    KeyRevoked,
    KidMismatch,
    RprInvalid,
    // Policy errors (E400-E499)
    TierInsufficient,
    PolicyViolation,
    DowngradeAttempt,
    CapabilityDenied,
    // Session errors (E500-E599)
    SessionNonceInvalid,
    SessionExpired,
    SessionReplay,
    SequenceOutOfOrder,
    // A2A errors (E600-E699)
    A2aHandshakeFailed,
    A2aBundleRejected,
    A2aCapabilityMismatch,
    A2aSessionClosed,
    A2aDecryptFailed,
    // Rate limiting (E700-E799)
    RateLimitExceeded,
    CircuitOpen,
    Backpressure,
    // Internal (E900-E999)
    InternalError,
    AlgorithmUnavailable,
    HsmError,
}

impl DcpErrorCode {
    /// The stable string identifier (e.g. `DCP-E100`).
    pub fn code(&self) -> &'static str {
        use DcpErrorCode::*;
        match self {
            BundleSchemaInvalid => "DCP-E001",
            ArtifactSchemaInvalid => "DCP-E002",
            VersionUnsupported => "DCP-E003",
            ManifestMissing => "DCP-E004",
            ClassicalSigInvalid => "DCP-E100",
            PqSigInvalid => "DCP-E101",
            CompositeBindingInvalid => "DCP-E102",
            SignatureMissing => "DCP-E103",
            SignatureExpired => "DCP-E104",
            HashChainBroken => "DCP-E200",
            ManifestHashMismatch => "DCP-E201",
            MerkleRootMismatch => "DCP-E202",
            DualHashMismatch => "DCP-E203",
            IntentHashMismatch => "DCP-E204",
            AgentRevoked => "DCP-E300",
            KeyExpired => "DCP-E301",
            KeyRevoked => "DCP-E302",
            KidMismatch => "DCP-E303",
            RprInvalid => "DCP-E304",
            TierInsufficient => "DCP-E400",
            PolicyViolation => "DCP-E401",
            DowngradeAttempt => "DCP-E402",
            CapabilityDenied => "DCP-E403",
            SessionNonceInvalid => "DCP-E500",
            SessionExpired => "DCP-E501",
            SessionReplay => "DCP-E502",
            SequenceOutOfOrder => "DCP-E503",
            A2aHandshakeFailed => "DCP-E600",
            A2aBundleRejected => "DCP-E601",
            A2aCapabilityMismatch => "DCP-E602",
            A2aSessionClosed => "DCP-E603",
            A2aDecryptFailed => "DCP-E604",
            RateLimitExceeded => "DCP-E700",
            CircuitOpen => "DCP-E701",
            Backpressure => "DCP-E702",
            InternalError => "DCP-E900",
            AlgorithmUnavailable => "DCP-E901",
            HsmError => "DCP-E902",
        }
    }

    /// Canonical message + retryable flag for this code.
    pub fn description(&self) -> (&'static str, bool) {
        use DcpErrorCode::*;
        match self {
            BundleSchemaInvalid => ("Bundle does not conform to DCP schema", false),
            ArtifactSchemaInvalid => ("Artifact does not conform to DCP schema", false),
            VersionUnsupported => ("DCP version not supported", false),
            ManifestMissing => ("Bundle manifest is missing", false),
            ClassicalSigInvalid => ("Classical (Ed25519) signature verification failed", false),
            PqSigInvalid => ("Post-quantum signature verification failed", false),
            CompositeBindingInvalid => ("Composite signature binding is invalid", false),
            SignatureMissing => ("Required signature is missing", false),
            SignatureExpired => ("Signature has expired", false),
            HashChainBroken => ("Audit hash chain integrity check failed", false),
            ManifestHashMismatch => ("Manifest hash does not match artifact", false),
            MerkleRootMismatch => ("Merkle root does not match audit entries", false),
            DualHashMismatch => ("Dual hash chain inconsistency detected", false),
            IntentHashMismatch => ("Intent hash does not match", false),
            AgentRevoked => ("Agent has been revoked", false),
            KeyExpired => ("Signing key has expired", false),
            KeyRevoked => ("Signing key has been revoked", false),
            KidMismatch => ("Key identifier does not match public key", false),
            RprInvalid => ("Responsible Principal Record is invalid", false),
            TierInsufficient => ("Security tier does not meet minimum requirement", false),
            PolicyViolation => ("Action violates policy", false),
            DowngradeAttempt => ("Security tier downgrade is not allowed", false),
            CapabilityDenied => ("Requested capability is not authorized", false),
            SessionNonceInvalid => ("Session nonce is invalid", false),
            SessionExpired => ("Session has expired", false),
            SessionReplay => ("Session replay detected", false),
            SequenceOutOfOrder => ("Message sequence out of order", false),
            A2aHandshakeFailed => ("A2A handshake failed", true),
            A2aBundleRejected => ("Peer rejected presented bundle", false),
            A2aCapabilityMismatch => ("Peer does not satisfy requested capabilities", false),
            A2aSessionClosed => ("A2A session is closed", false),
            A2aDecryptFailed => ("AES-GCM decryption failed (tag mismatch)", false),
            RateLimitExceeded => ("Rate limit exceeded", true),
            CircuitOpen => ("Circuit breaker is open", true),
            Backpressure => ("Backpressure applied — retry later", true),
            InternalError => ("Internal error", true),
            AlgorithmUnavailable => ("Requested algorithm is not registered", false),
            HsmError => ("Hardware security module reported an error", true),
        }
    }

    /// Is this error transient/retryable?
    pub fn retryable(&self) -> bool {
        self.description().1
    }
}

impl fmt::Display for DcpErrorCode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.code())
    }
}

/// Canonical DCP error value.
#[derive(Debug, Clone)]
pub struct DcpError {
    pub code: DcpErrorCode,
    pub message: String,
    pub retryable: bool,
    pub timestamp: String,
    pub details: serde_json::Value,
}

/// Build a `DcpError` with the standard message + retryable flag for the code.
pub fn create_dcp_error(
    code: DcpErrorCode,
    message: Option<String>,
    details: Option<serde_json::Value>,
) -> DcpError {
    let (default_msg, retryable) = code.description();
    DcpError {
        code,
        message: message.unwrap_or_else(|| default_msg.to_string()),
        retryable,
        timestamp: crate::v2::lifecycle::utc_now_iso_pub(),
        details: details.unwrap_or(serde_json::Value::Null),
    }
}

/// Wire-format sniff — returns `"json"` or `"cbor"`.
pub fn detect_wire_format(data: &[u8]) -> &'static str {
    if data.is_empty() {
        return "json";
    }
    match data[0] {
        0x7B | 0x5B | 0x20 | 0x0A | 0x0D | 0x09 => "json",
        _ => "cbor",
    }
}
